examples/io/legateio/legateio.py

```diff
@@ -0,0 +1,408 @@
+# from a dataset's header file.
+_CODES_TO_DTYPES = dict(
+    (context.type_system[t].code, t)
+    for t in (
+        ty.bool_,
+        ty.int8,
+        ty.int16,
+        ty.int32,
+        ty.int64,
+        ty.uint8,
+        ty.uint16,
+        ty.uint32,
+        ty.uint64,
+        ty.float16,
+        ty.float32,
+        ty.float64,
+        ty.complex64,
+        ty.complex128,
+    )
+)
+
+
+class IOArray:
+    """
+    A simple array implementation used in the tutorial.
+
+    This array can be passed to any Legate library call that supports Legate
+    data interface.
+    """
+
+    def __init__(self, store: Store, dtype: pa.lib.DataType) -> None:
+        self._store = store
+        self._dtype = dtype
+
+    @staticmethod
+    def from_legate_data_interface(data: dict[str, Any]) -> "IOArray":
+        """
+        Constructs an IOArray from a Legate data interface object.
+
+        The Legate data interface object contains its data under the
+        ``data`` field, which is a map from fields to arrays. Each array
+        is backed by one or more stores, and their layout follows Apache
+        arrow's data layout. For example, a nullable fixed type array
+        would be backed by two stores, one for the bitmask and one for
+        the actual array data.
+        """
+
+        assert data["version"] == 1
+        # For now, we assume that there's only one field in the container
+        field = next(iter(data["data"]))
+        stores = data["data"][field].stores()
+
+        # We only support non-nullable arrays
+        assert len(stores) == 2 and stores[0] is None
+        _, store = stores
+        return IOArray(store, field.type)
+
+    @property
+    def __legate_data_interface__(self) -> dict[str, Any]:
+        """
+        Constructs a Legate data interface object from this IOArray.
+        """
+
+        # Every IOArray would be mapped to a non-nullable Legate array
+        # of a fixed type
+        array = Array(self._dtype, [None, self._store])
+
+        # Create a field metadata to populate the data field
+        field = pa.field("Legate IO Array", self._dtype, nullable=False)
+
+        return {
+            "version": 1,
+            "data": {field: array},
+        }
+
+    def to_file(self, filename: str) -> None:
+        """
+        Dumps the IOArray to a single file.
+
+        Only a single task will be launched by this call.
+        """
+        task = context.create_auto_task(LegateIOOpCode.WRITE_FILE)
+
+        task.add_input(self._store)
+        task.add_scalar_arg(filename, ty.string)
+        # Request a broadcasting for the input. Since this is the only store
+        # argument to the task, Legate will launch a single task from this
+        # task descriptor.
+        task.add_broadcast(self._store)
+        task.execute()
+
+    def to_uneven_tiles(self, path: str) -> None:
+        """
+        Dumps the IOArray into (potentially) uneven tiles
+
```

Comment:
This does not run for me. It fails with:

```
Input shape: (5, 5)
[0 - 7f46b150f000]    0.850279 {3}{legateio}: Write to "test/.header"
[0 - 7f46b150f000]    0.850481 {3}{legateio}: Write a sub-array <0,0>..<4,4> to "test/0.0"
legion_python: /home/jwilke/Programs/legate-hello/legate.core/src/core/data/store.cc:300: legate::Domain legate::Store::domain() const: Assertion `!is_unbound_store_' failed.
Signal 6 received by node 0, process 1470359 (thread 7f46b150f000) - obtaining backtrace
Signal 6 received by process 1470359 (thread 7f46b150f000) at: stack trace: 20 frames
  [0] = /lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7f46c05f5090]
  [1] = /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb) [0x7f46c05f500b]
  [2] = /lib/x86_64-linux-gnu/libc.so.6(abort+0x12b) [0x7f46c05d4859]
  [3] = /lib/x86_64-linux-gnu/libc.so.6(+0x22729) [0x7f46c05d4729] 
```

Reply:
What compiler are you using? and what's the command for this issue? I can't reproduce it.